### ISSUE TITLE
git-info slow in large project

### ISSUE THREAD

**Issue by bergvandenp** [zimfw/zimfw#344]
`zmanage info`:

```
Zim commit ref:   d19c8dd
Zsh version:      zsh 5.7.1 (x86_64-apple-darwin19.0)
System info:      Darwin nvc3734.bla.local 19.0.0 Darwin Kernel Version 19.0.0: Thu Oct 17 16:17:15 PDT 2019; root:xnu-6153.41.3~29/RELEASE_X86_64 x86_64
```

- gist of .zimrc: https://gist.github.com/bergvandenp/b67b0de37d700d0a0fb8dcff923856a4
- gist of .zshrc: https://gist.github.com/bergvandenp/807412f9895e421956478a0f2b747482

Description
-----------
When using `steeef` prompt (and others) returning after an arbitrary command takes a couple of seconds when the current dir is in the root of a large git project.
I tested a couple of things with steeef settings and came to the conclusion that when I remove this line https://github.com/zimfw/zimfw/blob/d19c8dde68b338fcc096bbce683c47ad068b46d3/modules/prompt/themes/steeef.zsh-theme#L91, the waiting time is gone. Strangely editing the`ps1` line has no effect.
Fiddeling with the other git-info settings also don't effect. It really boils down to line 91.

FYI: I have no pending changes in the git project.

Steps to reproduce
------------------
```
cd <large project>
ls (and wait)
```

Images or other information
---------------------------
The problem occurred since the upgrade to Catalina + homebrew + more (that I don't remember exactly)

I got the same waiting times with `liquidprompt`, but it's harder to find the culprit in that (complex) config.

**Comment by ericbn** [comment]
Hi @bergvandenp, thanks for reporting this. It's documented in the [git-info README.md](https://github.com/zimfw/zimfw/blob/1ab110c264fc3da523b9ec288f84e60480f0bcd5/modules/git-info/README.md#verbose-mode), but maybe not being clean enough. Any suggestions on how to improve the documentation?

**Comment by bergvandenp** [comment]
Setting `zstyle ':zim:git-info' verbose no` does nothing in terms of my waiting time issue.
removing the line `zstyle ':zim:git-info:unindexed' format "${col_unidx}${ind_unidx}"` does make the difference. That's weird, right?

**Comment by ericbn** [comment]
Oh, that is weird, yeah!

Can you please temporarily edit your .zim/modules/git-info/functions/git-info as follows:
* Add to the top:
  ```
  PS4=$'%D{%s.%6.} '
  setopt xtrace
  ```
* Add to the bottom, before the "return 0" line:
  ```
  unsetopt xtrace
  ```

And then start a new shell session. Before each prompt, you will have an output of a timestamp and each instruction that is executed by git-info. Then can you please give us the output after executing yours steps to reproduce, both with and without the line 91 you mentioned in the steeef.zsh-theme file? I appreciate your help debugging this.

**Comment by bergvandenp** [comment]
Below the debug information. It seems that an extra `git diff-files` is executed without caching (marked yellow). The stuff marked blue is identical in both situations.
![issue](https://user-images.githubusercontent.com/209823/68393096-886a0780-016b-11ea-8161-d8c971ed77b8.png)

**Comment by bergvandenp** [comment]
looking at the sources `git-info verbose` is always `true`. How do I set this to false?

**Comment by bergvandenp** [comment]
the 'issue' here is that it hits my problem when verbose mode is set to false: https://github.com/zimfw/zimfw/blob/270b5d0390222c0a59b1fe7e4f7a2b96da7ced11/modules/git-info/functions/git-info#L129
The else branch is this is slow (obviously) with `git status --porcelain --ignore-submodules=${ignore_submodules}`

so when verbose is set to false AND the unindex format is defined --> my problem

**Comment by ericbn** [comment]
Verbose should enabled only when `zstyle ':zim:git-info' verbose 'yes'` is set. The line 129 of git-info that you mentioned is actually testing if verbose is not enabled:

```
❯ zstyle ':zim:git-info' verbose 'yes'                                                                                            1 ↵
❯ if ! zstyle -t ':zim:git-info' verbose; then print verbose disabled; else; print verbose enabled; fi
verbose enabled
❯ zstyle ':zim:git-info' verbose 'no'
❯ if ! zstyle -t ':zim:git-info' verbose; then print verbose disabled; else; print verbose enabled; fi
verbose disabled
```

Verbose should be enabled in the steeef prompt, as you can see in: https://github.com/zimfw/zimfw/blob/d19c8dde68b338fcc096bbce683c47ad068b46d3/modules/prompt/themes/steeef.zsh-theme#L87
But the xtrace output is the one for when verbose is disabled actually. Verbose uses `git status` instead of `git diff-files` and `git diff-index`.

The bottom line is:
* Why is the non-verbose code being executed for you when actually verbose is enabled in the steeef prompt? This is an issue for us.
* You're already getting the as-fast-as-we-have-now code because the non-verbose code is being executed (as noted in the [git-info README.md](https://github.com/zimfw/zimfw/blob/1ab110c264fc3da523b9ec288f84e60480f0bcd5/modules/git-info/README.md#verbose-mode)).
* The https://github.com/romkatv/gitstatus project is a "10x faster alternative to git status". There's a request to have a separate git-info module in zimfw that uses this gitstatus project instead of git directly (as we do now) in #336. I started working on it, but didn't finish it to be ready to push to master yet.

**Comment by bergvandenp** [comment]
Ok. Thanks for the investigation and the clear explanation! Should I close this issue?

**Comment by ericbn** [comment]
Before this is closed, I'm curious about the different timings. Can you please run these while in your repo:
```
git --version
time git diff-files --no-ext-diff --quiet --ignore-submodules=all
time git diff-index --no-ext-diff --quiet --cached --ignore-submodules=all HEAD
time git status --porcelain --ignore-submodules=all >/dev/null
```
I have `git version 2.21.0 (Apple Git-120)` and the timings of diff-files + diff-index > status (almost the double actually). And it used to be the case where diff-files + diff-index < status (much less, as I [got 3 years ago](https://gist.github.com/sindresorhus/3898739#gistcomment-1897703)). Looks like git changed in this sense.

**Comment by bergvandenp** [comment]
```
git version 2.21.0 (Apple Git-122.2)
git diff-files --no-ext-diff --quiet --ignore-submodules=all  0.03s user 11.09s system 822% cpu 1.352 total
git diff-index --no-ext-diff --quiet --cached --ignore-submodules=all HEAD  0.01s user 0.01s system 44% cpu 0.046 total
git status --porcelain --ignore-submodules=all > /dev/null  0.07s user 10.51s system 764% cpu 1.384 total
```

**Comment by ericbn** [comment]
Thanks! `git diff-files` used to be much faster [in the past](https://gist.github.com/sindresorhus/3898739#gistcomment-1897703). Of course, this also depends on the repository itself.

**Comment by ericbn** [comment]
UPDATE: Looks like `git diff-files` got way much slower in newer versions of git, while `git status` got faster.

First columns of times are what I had on [Oct 14, 2016](https://gist.github.com/sindresorhus/3898739#gistcomment-1897703) (didn't get the git version at that time, also not sure which Windows version I was using):

Second column of times is what I get today with *git version 2.21.0.windows.1* on the same repo and the same machine, currently with Windows 10 Home version 1903 build 18362.418:

Times as real time, in seconds.

| Command | Oct 14, 2016 |  Nov 7, 2019 |
| --- | --: | --: |
| `git ls-files --other --exclude-standard` | 2.756 | 3.150 |
| `git status --porcelain`  | 2.313 | 1.279|
| `git diff-index --no-ext-diff --quiet --cached HEAD` | 0.317 | 0.176 |
| `git diff-files --no-ext-diff --quiet`  | 0.022 | 0.936 |

@romkatv, if you allow me getting your attention, since you're a git status specialist by now.  :- )
Are you aware of `git diff-files` getting way much slower, while `git status` getting faster in newer versions of git? Any idea in which versions did these changes happen?

**Comment by romkatv** [comment]
Dunno. I always benchmark against `git status` as it's the fastest way to get all info.
